File: LeetCode/417. Pacific Atlantic Water Flow.py
# https://leetcode.com/problems/pacific-atlantic-water-flow/description/

from test import test


class Solution:
    def pacificAtlantic(self, heights: list[list[int]]) -> list[list[int]]:
        rows = len(heights)
        columns = len(heights[0])

        pacific: list[list[bool]] = [[False] * columns for _ in range(rows)]
        atlantic: list[list[bool]] = [[False] * columns for _ in range(rows)]

        directions: tuple[tuple[int, int], ...] = ((-1, 0), (1, 0), (0, -1), (0, 1))

        def dfs(y, x, ocean) -> None:
            if ocean[y][x]:
                return

            ocean[y][x] = True

            for delta_y, delta_x in directions:
                new_y = y + delta_y
                new_x = x + delta_x

                should_skip = (
                    new_y < 0
                    or new_y >= rows
                    or new_x < 0
                    or new_x >= columns
                    or ocean[new_y][new_x]
                )

                if should_skip:
                    continue

                if heights[new_y][new_x] < heights[y][x]:
                    continue

                dfs(new_y, new_x, ocean)

        for index in range(columns):
            dfs(0, index, pacific)
            dfs(rows - 1, index, atlantic)

        for index in range(max(rows, columns)):
            if index < rows:
                dfs(index, 0, pacific)
                dfs(index, columns - 1, atlantic)

            if index < columns:
                dfs(0, index, pacific)
                dfs(rows - 1, index, atlantic)

        result: list[list[int]] = []

        for y, x in [[y, x] for x in range(columns) for y in range(rows)]:
            if pacific[y][x] and atlantic[y][x]:
                result.append([y, x])

        return result


solution = Solution()


test(
    solution.pacificAtlantic,
    [
        {
            "input": [
                [
                    [1, 2, 2, 3, 5],
                    [3, 2, 3, 4, 4],
                    [2, 4, 5, 3, 1],
                    [6, 7, 1, 4, 5],
                    [5, 1, 1, 2, 4],
                ],
            ],
            "expected": [
                [0, 4],
                [1, 3],
                [1, 4],
                [2, 2],
                [3, 0],
                [3, 1],
                [4, 0],
            ],
        },
        {
            "input": [[[1]]],
            "expected": [[0, 0]],
        },
        {
            "input": [
                [
                    [10, 10, 10],
                    [10, 1, 10],
                    [10, 10, 10],
                ],
            ],
            "expected": [
                [0, 0],
                [0, 1],
                [0, 2],
                [1, 0],
                [1, 2],
                [2, 0],
                [2, 1],
                [2, 2],
            ],
        },
    ],
)

File: LeetCode/417. Pacific Atlantic Water Flow.js
// https://leetcode.com/problems/pacific-atlantic-water-flow/description/

const { test } = require('../test');

const pacificAtlantic = (heights) => {
  const rows = heights.length;
  const columns = heights[0].length;

  const pacific = Array.from({ length: rows }).map(() => Array.from({ length: columns }).fill(false));
  const atlantic = Array.from({ length: rows }).map(() => Array.from({ length: columns }).fill(false));

  const directions = [
    [1, 0],
    [-1, 0],
    [0, 1],
    [0, -1],
  ];

  const dfs = (y, x, ocean) => {
    if (ocean[y][x]) return;

    ocean[y][x] = true;

    for (const [deltaY, deltaX] of directions) {
      const newY = y + deltaY;
      const newX = x + deltaX;

      if (newY < 0 || newY >= rows || newX < 0 || newX >= columns || ocean[newY][newX]) {
        continue;
      }

      if (heights[newY][newX] < heights[y][x]) continue;

      dfs(newY, newX, ocean);
    }
  };

  for (let y = 0; y < rows; y++) {
    dfs(y, 0, pacific);
    dfs(y, columns - 1, atlantic);
  }

  for (let x = 0; x < columns; x++) {
    dfs(0, x, pacific);
    dfs(rows - 1, x, atlantic);
  }

  const result = [];

  for (let y = 0; y < rows; y++) {
    for (let x = 0; x < columns; x++) {
      if (pacific[y][x] && atlantic[y][x]) {
        result.push([y, x]);
      }
    }
  }

  return result;
};

test(pacificAtlantic, [
  {
    input: [
      [
        [1, 2, 2, 3, 5],
        [3, 2, 3, 4, 4],
        [2, 4, 5, 3, 1],
        [6, 7, 1, 4, 5],
        [5, 1, 1, 2, 4],
      ],
    ],
    expected: [
      [0, 4],
      [1, 3],
      [1, 4],
      [2, 2],
      [3, 0],
      [3, 1],
      [4, 0],
    ],
  },
  {
    input: [[[1]]],
    expected: [[0, 0]],
  },
  {
    input: [
      [
        [10, 10, 10],
        [10, 1, 10],
        [10, 10, 10],
      ],
    ],
    expected: [
      [0, 0],
      [0, 1],
      [0, 2],
      [1, 0],
      [1, 2],
      [2, 0],
      [2, 1],
      [2, 2],
    ],
  },
]);
